tools/nl/analysis/log/mine.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import csv
+from datetime import datetime
+from datetime import timedelta
+import json
+import logging
+
+from absl import app
+from absl import flags
+from google.cloud import bigtable
+from google.cloud.bigtable import row_filters
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('project', 'datcom-store', 'GCP project')
+flags.DEFINE_string('instance', 'website-data', 'BT instance')
+flags.DEFINE_string('table', 'nl-query', 'BT table')
+flags.DEFINE_integer('past_days', 5, 'Days to go back')
+flags.DEFINE_string('output_csv', '/tmp/nl_query_log.csv', 'Output CSV')
+
+_COL_SESSION = 'session_info'
+_COL_DATA = 'data'
+_COL_FEEDBACK = 'feedback'
+
+
+def _lidx(ctx, val, d, keys):
+  for k1, k2 in keys:
+    items = val.get(k1, [])
+    if d[k2] >= len(items):
+      logging.error(f'Invalid {ctx} {k1} {k2} {d[k2]} {len(items)}')
+      return {}
+    val = items[d[k2]]
+  return val
+
+
+def _midx(val, keys):
+  for i, k in enumerate(keys):
+    if i == len(keys) - 1:
+      val = val.get(k)
+    else:
+      val = val.get(k, {})
+  return val
+
+
+def _write_rows(key, data, csvw):
+  for fb in data[_COL_FEEDBACK]:
+    if 'queryId' in fb:
+      # Query-level feedback
+      item = _lidx(key, data[_COL_SESSION], fb, [('items', 'queryId')])
+      if not item:
+        continue
+      query = item['query']
+
+      ditem = _lidx(key, data, fb, [(_COL_DATA, 'queryId')])
+      if not ditem:
+        continue
+
+      config = {}
+
+    elif 'chartId' in fb:
+      cid = fb['chartId']
+      qitem = _lidx(key, data[_COL_SESSION], cid, [('items', 'queryIdx')])
+      if not qitem:
+        continue
+      query = qitem['query']
+
+      ditem = _lidx(key, data, cid, [(_COL_DATA, 'queryIdx')])
+      if not ditem:
+        continue
+
+      config = ditem['config']
+      config = _lidx(key, config, cid, [('categories', 'categoryIdx'),
+                                        ('blocks', 'blockIdx'),
+                                        ('columns', 'columnIdx'),
+                                        ('tiles', 'tileIdx')])
+      if not config:
+        continue
+
+    sentiment = fb['sentiment']
+
+    # Format xxx_yyyyyyyy#<project>
+    ts = int(key.split('#')[0].split('_')[1])
+    time = datetime.fromtimestamp(ts / 1000000.0).strftime('%Y-%m-%d %H:%M:%S')
+
+    output_row = {
```

Comment:
nit question: do we have access to the query url? if so, might be helpful to include it?

Reply:
great point.  added URL, also an absolute chart-index.